src/toolkits/object_detection/one_shot_object_detection/one_shot_object_detector.cpp

```diff
@@ -63,181 +62,200 @@ flex_dict build_annotation( ParameterSampler &parameter_sampler,
+
+  DASSERT_TRUE(background.get_image_data() != nullptr);
+  DASSERT_TRUE(starter.is_decoded());
+  DASSERT_TRUE(background.is_decoded());
+
+  boost::gil::rgba8_image_t::view_t starter_image_view =
+      create_starter_image_view(starter);
+
+  boost::gil::rgb8_image_t::view_t background_view = interleaved_view(
+      background.m_width, background.m_height,
+      (boost::gil::rgb8_pixel_t *)(background.get_image_data()),
+      background.m_channels * background.m_width  // row length in bytes
+  );
+  flex_image synthetic_image = create_synthetic_image(
+      starter_image_view, background_view, parameter_sampler, starter);
+  encode_image_inplace(synthetic_image);
+  return std::make_pair(synthetic_image, annotation);
+}
+
 gl_sframe augment_data(const gl_sframe &data,
-                       const std::string& image_column_name,
-                       const std::string& target_column_name,
-                       const gl_sarray &backgrounds,
-                       long long seed,
+                       const std::string &image_column_name,
+                       const std::string &target_column_name,
+                       const gl_sarray &backgrounds, long long seed,
                        bool verbose) {
   size_t backgrounds_size = backgrounds.size();
   size_t total_augmented_rows = data.size() * backgrounds_size;
-  table_printer table( { {"Images Augmented", 0}, {"Elapsed Time", 0}, {"Percent Complete", 0} } );
+  table_printer table(
+      {{"Images Augmented", 0}, {"Elapsed Time", 0}, {"Percent Complete", 0}});
   if (verbose) {
-    logprogress_stream << "Augmenting input images using " << backgrounds.size() << " background images." << std::endl;
+    logprogress_stream << "Augmenting input images using " << backgrounds_size
+                       << " background images." << std::endl;
     table.print_header();
   }
-
+  std::vector<std::string> output_column_names = {"image", "annotation"};
+  std::vector<flex_type_enum> output_column_types = {flex_type_enum::IMAGE,
+                                                     flex_type_enum::DICT};
+  gl_sframe_writer output_writer(output_column_names, output_column_types);
   auto column_index_map = generate_column_index_map(data.column_names());
-  std::vector<flexible_type> annotations;
-  std::vector<flexible_type> images;
-  int input_row_index = -1;
-  for (const auto& row: data.range_iterator()) {
-    input_row_index++;
-    flex_image object = row[column_index_map[image_column_name]].to<flex_image>();
-    std::string label = row[column_index_map[target_column_name]].to<flex_string>();
-    if (!(object.is_decoded())) {
-      decode_image_inplace(object);
-    }
-    int row_number = -1;
-    for (const auto& background_ft: backgrounds.range_iterator()) {
+  size_t image_column_index = column_index_map[image_column_name];
+  size_t target_column_index = column_index_map[target_column_name];
+  DASSERT_EQ(data[image_column_name].dtype(), flex_type_enum::IMAGE);
```

Comment:
If there's a chance these are user facing, they should be if(!(...) ){ log_and_throw(...); }. DASSERT_XXX is only for internal programmer consistency checks, not user-accessible code paths.

Reply:
Fixed.